services/codecov/codecov.service.js

```diff
@@ -100,40 +118,77 @@ module.exports = class Codecov extends BaseJsonService {
     }
   }
 
-  async fetch({ vcsName, user, repo, branch, token }) {
+  // Here for backward-compatibility purpose.
+  async legacyFetch({ vcsName, user, repo, branch, token }) {
     // Codecov Docs: https://docs.codecov.io/reference#section-get-a-single-repository
-    let url = `https://codecov.io/api/${vcsName}/${user}/${repo}`
-    if (branch) {
-      url += `/branches/${branch}`
-    }
-    const options = {}
-    if (token) {
-      options.headers = {
-        Authorization: `token ${token}`,
-      }
-    }
-    return this._requestJson({
-      schema,
-      options,
+    const url = `https://codecov.io/api/${vcsName}/${user}/${repo}${
+      branch ? `/branches/${branch}` : ''
+    }`
+    const { buffer } = await this._request({
       url,
+      options: {
+        headers: {
+          Accept: 'application/json',
+          Authorization: `token ${token}`,
+        },
+      },
       errorMessages: {
         401: 'not authorized to access repository',
         404: 'repository not found',
       },
     })
+    const json = parseJson(buffer)
+    return this.constructor._validate(json, legacySchema)
   }
 
-  transform({ json }) {
+  // Here for backward-compatibility purpose.
+  legacyTransform({ json }) {
     if (!json.commit || !json.commit.totals) {
       return { coverage: 'unknown' }
     }
 
     return { coverage: +json.commit.totals.c }
   }
 
-  async handle({ vcsName, user, repo, branch }, { token }) {
-    const json = await this.fetch({ vcsName, user, repo, branch, token })
-    const { coverage } = this.transform({ json })
+  // Doesn't support `flag` feature. Here for backward-compatibility purpose.
+  async legacyHandle({ vcsName, user, repo, branch }, { token }) {
+    const json = await this.legacyFetch({ vcsName, user, repo, branch, token })
+    const { coverage } = this.legacyTransform({ json })
+    return this.constructor.render({ coverage })
+  }
+
+  async fetch({ vcsName, user, repo, branch, token, flag }) {
+    const url = `https://codecov.io/${vcsName}/${user}/${repo}${
+      branch ? `/branches/${branch}` : ''
+    }/graph/badge.svg`
+    return this._requestSvg({
+      schema,
+      valueMatcher: svgValueMatcher,
+      url,
+      options: {
+        qs: { token, flag },
+      },
+      errorMessages: token ? { 400: 'invalid token pattern' } : {},
```

Comment:
The 400 will handle the scenario where someone tries to use an API token (such as an existing private badge) with a flag right?

Reply:
Yes, that's right.